connect/tls.go

```diff
@@ -326,27 +353,63 @@ func (cfg *dynamicTLSConfig) SetRoots(roots *x509.CertPool) error {
 	cfg.Lock()
 	defer cfg.Unlock()
 	cfg.roots = roots
-	cfg.notify()
+
+	if cfg.VerifyLeafWithRoots() == nil {
+		// only notify that we are ready once the roots can verify the leaf
+		cfg.notify()
+	}
+
+	// No error here even when verification fails. We cannot control
+	// the ordering of setting leaves and roots and therefore cannot
+	// make assumptions as to the correctness of the root or leaf when
+	// being set. The Ready method can be used to verify that the two
+	// go together.
 	return nil
 }
 
 // SetLeaf sets a new leaf.
 func (cfg *dynamicTLSConfig) SetLeaf(leaf *tls.Certificate) error {
 	cfg.Lock()
 	defer cfg.Unlock()
+	if err := parseLeafX509Cert(leaf); err != nil {
+		return err
+	}
 	cfg.leaf = leaf
-	cfg.notify()
+
+	if cfg.VerifyLeafWithRoots() == nil {
+		// only notify that we are ready once the roots can verify the leaf
+		cfg.notify()
+	}
+
+	// No error here even when verification fails. We cannot control
+	// the ordering of setting leaves and roots and therefore cannot
+	// make assumptions as to the correctness of the root or leaf when
+	// being set. The Ready method can be used to verify that the two
+	// go together.
 	return nil
 }
 
 // notify is called under lock during an update to check if we are now ready.
 func (cfg *dynamicTLSConfig) notify() {
-	if cfg.readyCh != nil && cfg.leaf != nil && cfg.roots != nil {
+	if cfg.readyCh != nil && cfg.leaf != nil && cfg.roots != nil && cfg.leaf.Leaf != nil {
```

Comment:
Does anyone feel strongly about keeping this check. Any place notify is currently called all these checks would have to have been done except for the readyCh != nil